store/tikv/kv.go

```diff
@@ -347,3 +395,60 @@ func (s *KVStore) SetTiKVClient(client Client) {
 func (s *KVStore) GetTiKVClient() (client Client) {
 	return s.client
 }
+
+func (s *KVStore) getMinResolveTSByStores(stores []*Store) uint64 {
+	failpoint.Inject("injectResolveTS", func(val failpoint.Value) {
+		injectTS := val.(int)
+		failpoint.Return(uint64(injectTS))
+	})
+	minSafeTS := uint64(math.MaxUint64)
+	s.resolveTSMu.RLock()
+	defer s.resolveTSMu.RUnlock()
+	for _, store := range stores {
+		safeTS := s.resolveTSMu.resolveTS[store.storeID]
+		if safeTS < minSafeTS {
+			minSafeTS = safeTS
+		}
+	}
+	return minSafeTS
+}
+
+func (s *KVStore) safeTSUpdater() {
+	t := time.NewTicker(time.Second * 2)
+	defer t.Stop()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for {
+		select {
+		case <-s.Closed():
+			return
+		case <-t.C:
+			s.updateResolveTS(ctx)
+		}
+	}
+}
+
+func (s *KVStore) updateResolveTS(ctx context.Context) {
+	stores := s.regionCache.getStoresByType(tikvrpc.TiKV)
+	tikvClient := s.GetTiKVClient()
+	wg := &sync.WaitGroup{}
+	wg.Add(len(stores))
+	for _, store := range stores {
+		go func(ctx context.Context, wg *sync.WaitGroup, store *Store) {
+			defer wg.Done()
+			// TODO: add metrics for updateSafeTS
+			resp, err := tikvClient.SendRequest(ctx, store.addr, tikvrpc.NewRequest(tikvrpc.CmdStoreSafeTS, &kvrpcpb.StoreSafeTSRequest{KeyRange: &kvrpcpb.KeyRange{
+				StartKey: []byte(""),
+				EndKey:   []byte(""),
+			}}), ReadTimeoutShort)
+			if err != nil {
+				return
+			}
+			safeTSResp := resp.Resp.(*kvrpcpb.StoreSafeTSResponse)
+			s.resolveTSMu.Lock()
+			s.resolveTSMu.resolveTS[store.storeID] = safeTSResp.GetSafeTs()
```

Comment:
What will happen if TiKV is an older version, which cannot provide safeTS?

Reply:
The deployment tool will upgrade tikv first before tidb, and PD will prevent the lower version tikv to join the cluster. Currently if the tikv couldn't provide safeTS for itself, the reslovedTS maintained in tidb will be zero, and the startTS of the txn will be fallbacked to the minStartTS.